### PR TITLE
vCPU allocation on add_vcpu (take 2)

### PR DESCRIPTION
**Pull request by glg-rv**
This PR adds support for vCPUs state to be donated from the host at vCPU creation, not at TVM creation.

Suggested reviewing is commit-by-commit. The preparation commits introduce various helpers. Then a few hardwired values are removed, and finally the patch that introduces the vcpu allocation.

Notes on the last patch:

1. VmCpus is now a fixed array of once, rather than a pagevec. This removes the need for variable allocation. The GuestVm structure increse from one to two 4k pages.
2. As opposed to the first patch, vm_cpu.rs is now much more similar to the baseline. VmCpusInner is called VmCpusEntry, and it is the bit that is allocated in a PageBox.
3. A vcpu is not present when the Once in the array hasn't been set.
4. num_vcpus() is gone, at least for now. Given the use we currently make, VmCpus should return an iterator (that can start from an arbitrary vcpu_id) for the present vcpus. This patch was becoming too long, and fixing things that were not part of the scope of this PR, so this is pushed to future work.

As a final note, I made about 20 different branches trying to get the type system to help us showing the relationship between status and vcpu state. There were a lot of subtle and not-so subtle problems. It can be done but it made this patch huge, and distracted from the mechanism that should be for review here. Definitely future work.

### CODE REVIEW

**Review by abrestic-rivos — COMMENTED** [review]
A few minor comments, but overall LGTM.

File: src/guest_tracking.rs [hunk 1]
```diff
@@ -91,8 +91,15 @@ impl<T: GuestStagePagingMode> Clone for GuestVm<T> {
...
 impl<T: GuestStagePagingMode> GuestVm<T> {
+    /// Return required pages necessary to create a GuestVM.
+    pub const fn required_pages() -> u64 {
+        PageArc::<RwLock<GuestVmInner<T>>>::required_pages()
+    }
+
     /// Creates a new initializing `GuestVm` from `vm`, using `page` as storage.
     pub fn new(vm: Vm<T>, page: Page<InternalClean>) -> Self {
+        // assert it fits in a single page for now.
+        assert!(Self::required_pages() == 1);
```
[the remark this answers is not in the document]
Comment (glg-rv) [reply src/guest_tracking.rs]: This was a line that was added to let the preparation commit compile in case of bisect, but it is immediately overwritten. Didn't know `const_assert` though. Thanks!

File: riscv-pages/src/sequential_pages.rs [hunk 2]
```diff
@@ -236,6 +243,44 @@ impl<S: State, I: Iterator<Item = Page<S>>> fmt::Debug for Error<S, I> {
...
+            .checked_add_pages_with_size(count, page_size)
+            .unwrap();
+        self.pages.count -= count;
+        // Safe because `self` owns the memory, and we're creating a sequence from pages that
+        // have been removed.
+        unsafe { Some(SequentialPages::from_mem_range(addr, page_size, count).unwrap()) }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let count = ((self.pages.count + self.n - 1) / self.n) as usize;
```
Comment (atulkharerivos) [inline riscv-pages/src/sequential_pages.rs:277]: Should we check for division by zero (self.n == 0) here?

File: src/vm.rs [hunk 3]
```diff
@@ -27,7 +27,7 @@ use crate::vm_cpu::{
 use crate::vm_pages::Error as VmPagesError;
 use crate::vm_pages::{
     ActiveVmPages, AnyVmPages, InstructionFetchError, PageFaultType, VmPages, VmPagesRef,
-    VmRegionList, TVM_REGION_LIST_PAGES, TVM_STATE_PAGES,
+    VmRegionList, TVM_REGION_LIST_PAGES,
```
Comment (atulkharerivos) [inline src/vm.rs:31]: Unrelated nit: In the commit title, "pages numbers" should be "page numbers".

File: src/vm_pages.rs [hunk 4]
```diff
@@ -68,10 +70,6 @@ pub type InstructionFetchResult = core::result::Result<DecodedInstruction, Instr
 /// The number of pages for the `VmRegionList` vector.
 pub const TVM_REGION_LIST_PAGES: u64 = 1;
```
[the remark this answers is not in the document]
Comment (glg-rv) [reply src/vm_pages.rs:71]: Yes, could be a good idea. The advantage of keeping in a page vec is that it automatically resize on the number of pages. The disadvantage is that we don't control the precise number of entries we have. Which can be a problem especially if the structure grows.
I think that now that we have multiple pages backing for the basic structures of salus, a rethought is in order.

File: sbi/src/tee_host.rs [hunk 5]
```diff
@@ -454,7 +450,9 @@ pub enum TeeHostFunction {
         guest_id: u64,
         /// a1 = vCPU id
         vcpu_id: u64,
-        /// a2 = page address of shared state structure
+        /// a2 = address of the first page donated for the vCPU state
+        state_addr: u64,
```
[the remark this answers is not in the document]
Comment (glg-rv) [reply sbi/src/tee_host.rs:454]: Oh yes.

File: src/guest_tracking.rs [hunk 6]
```diff
@@ -96,18 +96,16 @@ impl<T: GuestStagePagingMode> GuestVm<T> {
...
 
-    /// Creates a new initializing `GuestVm` from `vm`, using `page` as storage.
-    pub fn new(vm: Vm<T>, page: Page<InternalClean>) -> Self {
-        // assert it fits in a single page for now.
-        assert!(Self::required_pages() == 1);
+    /// Creates a new initializing `GuestVm` from `vm`, using `pages`
+    /// as storage. Caller must ensure that `pages` contains
+    /// contiguous pages that can hold the PageArc.
+    pub fn new(vm: Vm<T>, pages: SequentialPages<InternalClean>) -> Self {
+        assert!(Self::required_pages() <= pages.len());
```
[the remark this answers is not in the document]
Comment (glg-rv) [reply src/guest_tracking.rs:103]: Hmm. I can return the Sequential Pages back in the error, but somehow it gets dangerous because a simple '?' might resul t in leaked pages. I have become assert happy when handling SequentialPages.
I think the idea was to fail on the area of salus where we were passing around SequentialPages.

File: riscv-pages/src/sequential_pages.rs (hunk 2 above)
Comment (glg-rv) [reply riscv-pages/src/sequential_pages.rs:277]: Good catch, as the n can be the result of some dynamic calculation and not the static use we do now. Will prefent chunks of zero (that would be an infinite loop) to be created.

File: src/vm_cpu.rs [hunk 7]
```diff
@@ -1089,86 +1084,89 @@ impl<'vcpu> IdleVmCpu<'vcpu> {
...
-        *status = VmCpuStatus::PoweredOff;
-        Ok(())
     }
 
     /// Returns a reference to the vCPU with `vcpu_id` if it exists and is not currently running.
     /// The returned `IdleVmCpu` is guaranteed not to change state until it is dropped.
     pub fn get_vcpu(&self, vcpu_id: u64) -> Result<IdleVmCpu> {
-        let entry = self.inner.get(vcpu_id as usize).ok_or(Error::BadCpuId)?;
+        let once_entry = self.inner.get(vcpu_id as usize).ok_or(Error::BadCpuId)?;
+        let entry = once_entry.get().ok_or(Error::VmCpuNotFound)?;
```
[the remark this answers is not in the document]
Comment (glg-rv) [reply src/vm_cpu.rs:1154]: Agreed, any vcpu_id that doesn't correspond to a vcpu is equally bad and non present for us. Will fix accordingly.

File: src/guest_tracking.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (glg-rv) [reply src/guest_tracking.rs:103]: I'll think about this in a next PR. I think if this change should change for all interface that now are dependent on SequentialPages.

File: src/vm_cpu.rs [hunk 8]
```diff
@@ -1021,11 +1015,12 @@ impl<T: GuestStagePagingMode> Drop for ActiveVmCpu<'_, '_, '_, T> {
         }
 
         // Return this vCPU to the `VmCpus` container.
-        let entry = self
+        let once_entry = self
             .container
             .inner
             .get(self.vcpu.vcpu_id as usize)
             .unwrap();
+        let entry = once_entry.get().unwrap();
```
[the remark this answers is not in the document]
Comment (glg-rv) [reply src/vm_cpu.rs:1023]: Done.

File: src/vm_cpu.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (glg-rv) [reply src/vm_cpu.rs:1154]: Done.

File: src/vm_pages.rs [hunk 9]
```diff
@@ -815,6 +815,11 @@ impl<T: GuestStagePagingMode> VmPages<T> {
     pub const fn required_state_pages() -> u64 {
         GuestVm::<T>::required_pages() + TVM_REGION_LIST_PAGES
     }
+
+    /// Returns the number of pages that must be donated to create a vCPU.
+    pub const fn vcpu_required_state_pages() -> u64 {
+        PageBox::<VmCpusEntry>::required_pages()
+    }
```
[the remark this answers is not in the document]
Comment (glg-rv) [reply src/vm_pages.rs:822]: Moved.

File: src/vm_pages.rs [hunk 10]
```diff
@@ -1151,22 +1164,45 @@ impl<'a, T: GuestStagePagingMode> FinalizedVmPages<'a, T> {
             GuestStagePageTable::new(guest_root_pages, id, self.inner.page_tracker.clone())
                 .unwrap();
 
-        let mut state_pages = self.assign_state_pages_for(state_pages, id);
-        let box_page = state_pages.next().unwrap();
-        let region_vec_pages = SequentialPages::from_pages(state_pages).unwrap();
+        let region_vec_pages =
+            SequentialPages::from_pages(self.assign_state_pages_for(region_vec_pages, id)).unwrap();
```
[the remark this answers is not in the document]
Comment (glg-rv) [reply src/vm_pages.rs:1168]: Done.

File: src/vm_pages.rs [hunk 11]
```diff
@@ -1151,22 +1164,45 @@ impl<'a, T: GuestStagePagingMode> FinalizedVmPages<'a, T> {
...
+        if !converted_pages.is_contiguous() {
+            return Err(Error::NonContiguousPages);
+        }
+
+        let guest_id = to.page_owner_id();
+        let vcpu_entry = VmCpusEntry::new(vcpu_id, guest_id, shared_area);
+
+        let vcpu_pages =
+            SequentialPages::from_pages(self.assign_state_pages_for(converted_pages, guest_id))
+                .unwrap();
```
[the remark this answers is not in the document]
Comment (glg-rv) [reply src/vm_pages.rs:1203]: Done.

File: riscv-pages/src/sequential_pages.rs (hunk 2 above)
Comment (glg-rv) [reply riscv-pages/src/sequential_pages.rs:277]: Done.

File: src/vm.rs (hunk 3 above)
Comment (glg-rv) [reply src/vm.rs:31]: Fixed. :)

**Comment by glg-rv** [comment]
Updated with review comments. Diff from previous is here: https://github.com/rivosinc/salus/compare/86dd8f6a16222b18fc05a1063157feade193e766..3d10d88e09ebbbe3c49db21f1631582eabd5f735

File: src/guest_tracking.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (dgreid) [reply src/guest_tracking.rs:103]: See `SequentialPages::from_pages`. It does something similar returning the passed pages back in the error case. Having the pages be part of the error type prevents them from being leaked, a `?` can't generate an error without the pages.

File: src/guest_tracking.rs [hunk 12]
```diff
@@ -95,7 +95,11 @@ impl<T: GuestStagePagingMode> GuestVm<T> {
     pub fn new(vm: Vm<T>, page: Page<InternalClean>) -> Self {
         let page_tracker = vm.page_tracker();
         Self {
-            inner: PageArc::new_with(RwLock::new(GuestVmInner::new(vm)), page, page_tracker),
+            inner: PageArc::new_with(
+                RwLock::new(GuestVmInner::new(vm)),
+                SequentialPages::from_pages([page]).unwrap(),
```
Comment (dgreid) [inline src/guest_tracking.rs]: unwrap isn't needed here: `SequentialPages::from(page)` is infallible.

Comment (glg-rv) [reply src/guest_tracking.rs]: Sure, this was temporary code of a preparation step. Nothing we'd ever run. Mostly show of future usage.

File: src/guest_tracking.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (glg-rv) [reply src/guest_tracking.rs:103]: Fair enough. So the strategy is having an error so annoying to deal with that you have to put unwrap, and explain why. It's not a bad one. :)

**Review by abrestic-rivos — APPROVED** [review]
LGTM, thanks.

**Review by atulkharerivos — APPROVED** [review]

**Comment by glg-rv** [comment]
Pushed to change the intermediate commit with Dylan's comments. Rebase makes the compare tricky.

**Review by dgreid — COMMENTED** [review]
Thanks, this looks good. Just a couple of nits.

File: riscv-pages/src/sequential_pages.rs [hunk 13]
```diff
@@ -236,6 +244,45 @@ impl<S: State, I: Iterator<Item = Page<S>>> fmt::Debug for Error<S, I> {
...
+        if self.pages.count == 0 {
+            return None;
+        }
+        let addr = self.pages.addr;
+        let page_size = self.pages.page_size;
+        let count = if self.pages.count > self.n.get() {
+            self.n.get()
+        } else {
+            self.pages.count
+        };
```
[the remark this answers is not in the document]
Comment (glg-rv) [reply riscv-pages/src/sequential_pages.rs:265]: I actually much prefer in general an explicit if to a function like that. But then, 🤷🏼‍♂️.

File: sbi/src/api/tee_host.rs [hunk 14]
```diff
@@ -160,19 +146,28 @@ pub fn get_vcpu_register_set(vmid: u64, index: u64) -> Result<RegisterSetLocatio
...
 /// # Safety
 ///
 /// The caller must own the pages referenced by `shared_page_addr`, for the number of pages
 /// sufficient to hold the structure layout returned from `get_vcpu_mem_layout()`. Since memory
 /// within the shared-memory state area may be read or written by the TSM at any time, the caller
 /// must treat the memory as volatile for the lifetime of the TVM.
-pub unsafe fn add_vcpu(vmid: u64, vcpu_id: u64, shared_page_addr: u64) -> Result<()> {
+///
+/// The address `state_page_addr` must reference confidential memory pages. The caller must provide
+/// `TsmInfo::tvm_vcpu_state_pages` pages.
```
[the remark this answers is not in the document]
Comment (glg-rv) [reply sbi/src/api/tee_host.rs:160]: Yes.

If you follow that code on the other side of Salus you'll find yourself in [`guest_add_vcpu`](https://github.com/rivosinc/salus/blob/main/src/vm.rs#L1210). 
From there, we arrive to [`create_guest_vm`](https://github.com/glg-rv/salus/blob/topic/vcpu_allocation/src/vm_pages.rs#L1117). When we arrive here, the only check we've done in the address passed is essentially that it is page aligned, something that happened when we annotated the raw address into a full `GuestPageAddr`.

Here is where things actually happen -- we're in `vm_pages.rs` so that's expected. You should notice that the function is divided in two parts. The first is where we try to get ranges of contiguous converted pages and exit with error early in case we don't.
And the other where we take these pages, assign them, and make them into SequentialPages to create data structures. Anything that handles SequentialPages can't fail.
This is due to the fact that if we fail _before_ the assign+SequentialPages, we would still be able to release those pages on variable drop. Once we assign them, or they pages are saved in a SequentialPages, we won't.

This is why we check early the memory requirements, and are the memory requirement you're asking about.

The fundamental check consist in `get_converted_pages`, that gets a range of memory that have been previously been converted. This is the check that it is confidential, because pages would have followed the protocol as described in `TsmConvertPages` (and the fence procedure of TLB invalidation).
This will also enforce the precise number of pages we'll receive (or return an error).

So it does error out if it isn't confidential or it's not big enough, and does not write anything in host memory in those cases.

File: riscv-pages/src/sequential_pages.rs [hunk 15]
```diff
@@ -236,6 +244,45 @@ impl<S: State, I: Iterator<Item = Page<S>>> fmt::Debug for Error<S, I> {
...
+        self.pages.addr = self
+            .pages
+            .addr
+            .checked_add_pages_with_size(count, page_size)
+            .unwrap();
```
[the remark this answers is not in the document]
Comment (dgreid) [reply riscv-pages/src/sequential_pages.rs:270]: sure, but it seems safe to move the count decrement before the addr increment and skip the addr increment if count is zero, right?

Also sorry, it's probably my bug from the other iterator being carried over:(

File: sbi/src/api/tee_host.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (dgreid) [reply sbi/src/api/tee_host.rs:160]: Thanks for that detailed evaluation. In that case, I'd move the 'state_page_addr' restriction in the comment to above the safety section as it isn't unsafe, from rust's perspective, to pass an arbitrary pointer to this function.

File: riscv-pages/src/sequential_pages.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (glg-rv) [reply riscv-pages/src/sequential_pages.rs:270]: Yes this is how I am modifying the function atm. Am also adding tests for all corner cases.

File: sbi/src/api/tee_host.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (glg-rv) [reply sbi/src/api/tee_host.rs:160]: Ah yes I will. I think in some past iteration was told to add the comment there. I didn't know what it was meant for the safety section there. Just doing what I was told. :-)

**Comment by glg-rv** [comment]
Reworked the chunk allocator and added tests. Comment changes on api/tee_host.

**Review by dgreid — APPROVED** [review]
